cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java

```diff
@@ -126,45 +155,121 @@
     this.streamCoordinatorClient = streamCoordinatorClient;
     this.privilegesManager = privilegesManager;
     this.authorizationEnforcer = authorizationEnforcer;
+    // TODO make maximum size and expire after write configurable?
+    this.appInjectors = CacheBuilder.newBuilder()
+      .maximumSize(10)
+      .expireAfterWrite(15, TimeUnit.MINUTES)
+      .removalListener(new RemovalListener<ApplicationId, Injector>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<ApplicationId, Injector> notification) {
+          Injector injector = notification.getValue();
+          if (injector != null) {
+            PreviewRuntimeService service = injector.getInstance(PreviewRuntimeService.class);
+            service.stopAndWait();
+          }
+          ApplicationId application = notification.getKey();
+          if (application == null) {
+            return;
+          }
+          java.nio.file.Path previewDirPath = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                                                        "preview", application.getApplication()).toAbsolutePath();
+
+          try {
+            Files.deleteIfExists(previewDirPath);
+          } catch (IOException e) {
+            LOG.warn("Error deleting the preview directory {}", previewDirPath, e);
+          }
+        }
+      })
+      .build();
   }
 
   @POST
   @Path("/previews")
   public void start(HttpRequest request, HttpResponder responder,
                     @PathParam("namespace-id") String namespaceId) throws Exception {
+    NamespaceId namespace = new NamespaceId(namespaceId);
+    ApplicationId application = namespace.app(RunIds.generate().getId());
+
+    AppRequest appRequest;
+    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8)) {
+      appRequest = GSON.fromJson(reader, AppRequest.class);
+    } catch (IOException e) {
```

Comment:
maybe just remove this catch block. ``HttpExceptionHandler`` will log an error and report the message.